wandb/sdk/wandb_run.py

```diff
@@ -2787,6 +2341,537 @@ def mark_preempting(self) -> None:
+
+        if not settings._offline and exit_code:
+            info.append("Press ctrl-c to abort syncing.")
+
+        printer.display(f'{" ".join(info)}')
+
+    # fixme: Temporary hack until we move to rich which allows multiple spinners
+    @staticmethod
+    def _footer_file_pusher_status_info(
+        poll_exit_responses: Optional[
+            Union[PollExitResponse, Dict[str, Optional[PollExitResponse]]]
+        ] = None,
+        *,
+        printer: Union["PrinterTerm", "PrinterJupyter"],
+    ) -> None:
+        if not poll_exit_responses:
+            return
+        if isinstance(poll_exit_responses, PollExitResponse):
+            Run._footer_single_run_file_pusher_status_info(
+                poll_exit_responses, printer=printer
+            )
+        elif isinstance(poll_exit_responses, dict):
+            poll_exit_responses_list = [
+                response for response in poll_exit_responses.values()
+            ]
+            assert all(
+                isinstance(response, (PollExitResponse, None))  # type: ignore
+                for response in poll_exit_responses_list
+            )
+            if len(poll_exit_responses_list) == 0:
+                return
+            elif len(poll_exit_responses_list) == 1:
+                Run._footer_single_run_file_pusher_status_info(
+                    poll_exit_responses_list[0], printer=printer
+                )
+            else:
+                Run._footer_multiple_runs_file_pusher_status_info(
+                    poll_exit_responses_list, printer=printer
+                )
+        else:
+            raise ValueError(
+                f"got the type {type(poll_exit_responses)} for poll_exit_responses expected either None, PollExitResponse or a Dict[str, Union[PollExitResponse, None]]"
+            )
+
+    @staticmethod
+    def _footer_single_run_file_pusher_status_info(
+        pool_exit_response: Optional[PollExitResponse] = None,
+        *,
+        printer: Union["PrinterTerm", "PrinterJupyter"],
+    ) -> None:
+        # todo: is this same as settings._offline?
+        if not pool_exit_response:
+            return
+
+        progress = pool_exit_response.pusher_stats
+        done = pool_exit_response.done
+
+        megabyte = wandb.util.POW_2_BYTES[2][1]
+        line = f"{progress.uploaded_bytes/megabyte :.2f} MB of {progress.total_bytes/megabyte:.2f} MB uploaded ({progress.deduped_bytes/megabyte:.2f} MB deduped)\r"
+
+        percent_done = (
+            1.0
+            if progress.total_bytes == 0
+            else progress.uploaded_bytes / progress.total_bytes
+        )
+
+        printer.progress_update(line, percent_done)
+        if done:
+            printer.progress_close()
+
+            dedupe_fraction = (
+                progress.deduped_bytes / float(progress.total_bytes)
+                if progress.total_bytes > 0
+                else 0
+            )
+            if dedupe_fraction > 0.01:
+                printer.display(
```

Comment:
Does the codecov complaint means we don't test this logic anywhere? (i mean large dedup fractions)

Reply:
yeah probably the case